pkg/store/proxy.go

```diff
@@ -179,40 +194,61 @@ type streamSeriesSet struct {
 
 	currSeries *storepb.Series
 	recvCh     chan *storepb.Series
+
+	errMtx sync.Mutex
+	err error
+
+	name string
 }
 
 func startStreamSeriesSet(
+	ctx context.Context,
+	wg *sync.WaitGroup,
 	stream storepb.Store_SeriesClient,
 	warnCh chan<- *storepb.SeriesResponse,
-	bufferSize int,
+	name string,
+	partialResponse bool,
 ) *streamSeriesSet {
 	s := &streamSeriesSet{
 		stream: stream,
 		warnCh: warnCh,
-		recvCh: make(chan *storepb.Series, bufferSize),
+		recvCh: make(chan *storepb.Series, 10),
+		name: name,
 	}
-	go s.fetchLoop()
-	return s
-}
 
-func (s *streamSeriesSet) fetchLoop() {
-	defer close(s.recvCh)
-	for {
-		r, err := s.stream.Recv()
-		if err == io.EOF {
-			return
-		}
-		if err != nil {
-			s.warnCh <- storepb.NewWarnSeriesResponse(errors.Wrap(err, "receive series"))
-			return
-		}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer close(s.recvCh)
+		for {
+			r, err := s.stream.Recv()
+			if err == io.EOF {
+				return
+			}
 
-		if w := r.GetWarning(); w != "" {
-			s.warnCh <- storepb.NewWarnSeriesResponse(errors.New(w))
-			continue
+			if err != nil {
+				if partialResponse {
+					s.warnCh <- storepb.NewWarnSeriesResponse(errors.Wrap(err, "receive series"))
+					return
+				}
+				if ctx.Err() != nil {
```

Comment:
We pass same context to store server connections, so If the context is cancelled stream will return error anyway. So there is no need. But I agree this might be not straightforward